rand_core/src/impls.rs

```diff
@@ -335,6 +341,171 @@ impl<R: BlockRngCore + SeedableRng> SeedableRng for BlockRng<R> {
     }
 }
 
+
+
+/// Wrapper around PRNGs that implement [`BlockRngCore`] to keep a results
+/// buffer and offer the methods from [`RngCore`].
+///
+/// This is similar to [`BlockRng`], but specialized for algorithms that operate
+/// on `u64` values.
+///
+/// [`BlockRngCore`]: ../BlockRngCore.t.html
+/// [`RngCore`]: ../RngCore.t.html
+/// [`BlockRng`]: struct.BlockRng.html
+#[derive(Clone)]
+#[cfg_attr(feature="serde-1", derive(Serialize, Deserialize))]
+pub struct BlockRng64<R: BlockRngCore + ?Sized> {
+    #[cfg_attr(feature="serde-1", serde(bound(
+        serialize = "R::Results: Serialize",
+        deserialize = "R::Results: Deserialize<'de>")))]
+    pub results: R::Results,
+    pub index: usize,
+    pub half_used: bool, // true if only half of the previous result is used
```

Comment:
I think `fill_bytes` should reset `half_used = false` ? In fact I think the x86 version could also use the left-over bytes easily (because LE means the bytes are contiguous); the other version could also but would need an extra copy.

Reply:
I don't think tests could catch this anyway; there's no memory unsafety, just the possibility of skipping one word of output and using another twice.

Reply:
Oops, I though we tested for it in `test_isaac64_true_values_mixed`, but that tests only mixing `next_u64` and `next_u32`.